File: models_matlab/model_RDQ20_MF.m
function output = model_RDQ20_MF(input, params)
%__________________________________________________________________________
%
% This code implements the mean-field ODE model (MF-ODE) for cardiomyocytes
% force generation presented in [1, 2].
%
% Inputs:
%  - input: struct containing the input data
%     - input.times: time instants [s]
%     - input.Ca:    intracellular calcium ions concentration [micro M]
%     - input.SL:    sarcomere length [micro m]
%  - params: struct containing the model parameters
%
% Outputs:
%  - output: struct containing the output data
%     - output.times: time instants [s]
%     - output.Ca:    intracellular calcium ions concentration [micro M]
%     - output.SL:    sarcomere length [micro m]
%     - output.P:     permissivity [-]
%     - output.Ta:    active tension [kPa]
%
%__________________________________________________________________________
%
% Author: Francesco Regazzoni - MOX, Politecnico di Milano
% Email:  francesco.regazzoni@polimi.it
% Date:   2020
%__________________________________________________________________________
%
% References:
%
% [1] F. Regazzoni "Mathematical modeling and Machine Learning for the
%     numerical simulation of cardiac electromechanics", PhD Thesis -
%     Politecnico di Milano (2020)
%     http://hdl.handle.net/10589/152617
% [2] F. Regazzoni, L. Dede', A. Quarteroni "Biophysically detailed
%     mathematical models of multiscale cardiac active mechanics",
%     PLOS Computational Biology (2020)
%     https://doi.org/10.1371/journal.pcbi.1008294
%
%__________________________________________________________________________

    %% Model parameters
    if nargin == 1
        params = '../params/params_RDQ20-MF_human_body-temperature.json';
    end
    if ischar(params)
        params = jsondecode(fileread(params));
    end
    LA       = params.geometry.LA;             % [micro m]
    LM       = params.geometry.LM;             % [micro m]
    LB       = params.geometry.LB;             % [micro m]
    SL0      = params.geometry.SL0;            % [micro m]
    Q        = params.RU_steady_state.Q;       % [-]
    Kd0      = params.RU_steady_state.Kd0;     % [micro M]
    alphaKd  = params.RU_steady_state.alphaKd; % [micro M / micro m]
    mu       = params.RU_steady_state.mu;      % [-]
    gamma    = params.RU_steady_state.gamma;   % [-]
    Koff     = params.RU_kinetics.Koff;        % [s^-1]
    Kbasic   = params.RU_kinetics.Kbasic;      % [s^-1]
    r0       = params.XB_cycling.r0;           % [s^-1]
    alpha    = params.XB_cycling.alpha;        % [-]
    mu0_fP   = params.XB_cycling.mu0_fP;       % [s^-1]
    mu1_fP   = params.XB_cycling.mu1_fP;       % [s^-1]
    a_XB     = params.upscaling.a_XB;          % [kPa]

    %% Numerical parameters
    dt_RU = 2.5e-5;                                % [s]
    freqXB = round(1e-3 / dt_RU);                  % [-]
    freqenceRatesUpdate = round(2.5e-4 / 2.5e-5);  % [-]

    %% Initialization
    LMh = (LM - LB) / 2;
    frac_SO =  @(SL)((SL>LA)     .*(SL<=LM)     .*(SL-LA) + ...
                     (SL>LM)     .*(SL<=2*LA-LB).*(SL+LM-2*LA)*.5 + ...
                     (SL>2*LA-LB).*(SL<=2*LA+LB).*LMh + ...
                     (SL>2*LA+LB).*(SL<=2*LA+LM).*(LM+2*LA-SL)*.5 ...
               )/LMh;
    Kd = @(SL) Kd0 - alphaKd*(2.15-SL);
    kC = zeros(2,2);     %kC(A,a) = P( C_i^t+dt =~ A | (C_i,T_i)^t = (A,a) ) / dt
    kT = zeros(2,2,2,2); %kT(a,b,c,B) = P( T_i^t+dt =~ b | (T_i-1,T_i,T_i+1,C_i)^t = (a,b,c,B) ) / dt
    expMat = permute([0 1; 1 2],[1,3,2]);
    kC(2,1) = Koff;
    kC(2,2) = Koff/mu;
    kT(:,2,:,1) = Kbasic*gamma.^(2-expMat);
    kT(:,2,:,2) = Kbasic*gamma.^(2-expMat);
    kT(:,1,:,1) = Q*Kbasic/mu*gamma.^expMat;
    kT(:,1,:,2) = Q*Kbasic*gamma.^expMat;

    %% Time loop
    times = min(input.times):dt_RU:max(input.times);
    Ca    = interp1(input.times,input.Ca,times);
    SL    = interp1(input.times,input.SL,times);
    nT    = length(times);
    P     = zeros(1,nT);
    Ta    = zeros(1,nT);

    time_init = tic();
    fprintf('RDQ20-MF model. Computing... ')
    for iT = 1:nT
        if mod(iT-1,freqenceRatesUpdate) == 0
            update_rates_RU(Ca(iT),SL(iT));
        end
        if iT == 1
            x_RU = zeros(2,2,2,2); %x_RU(a,b,c,B) = P( (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) )
            x_RU(1,1,1,1) = 1;
            x_XB = zeros(2,2);
        else
            x_RU = x_RU + dt_RU*RU_rhs(x_RU);
            if mod(iT-1,freqXB) == 0
                dSLdt = (SL(iT)-SL(iT-freqXB))/(dt_RU*freqXB);
                x_XB = XB_advance(x_RU,x_XB,dt_RU*freqXB,dSLdt);
            end
        end
        P(iT) = sum(reshape(x_RU(:,2,:),[],1))*frac_SO(SL(iT));
        Ta(iT) = a_XB*sum(x_XB(2,:))*frac_SO(SL(iT));
    end
    fprintf('done. Time elapsed: %1.3f s\n', toc(time_init))

    output.t  = times';
    output.Ca = Ca';
    output.SL = SL';
    output.P  = P';
    output.Ta = Ta';

    %% Functions definition
    function update_rates_RU(Ca,SL)
        Kon = Koff/Kd(SL); % [micro M^-1 * s^-1]
        kC(1,:) = Kon*Ca;
    end

    function rhs = RU_rhs(x_RU)
        PhiT_C = x_RU.*kT;                    % ...(a,b,c,B) =  P( T_i^t+dt =~ b     , (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        kT_L = permute(sum(sum(PhiT_C,1),4)./sum(sum(x_RU,1),4),[2,3,1,4]);  % ...(a,b,_,_) ~= P( T_{i-1}^t+dt =~ a | (T_{i-1},T_i)^t = (a,b) ) / dt
        kT_R = permute(sum(sum(PhiT_C,3),4)./sum(sum(x_RU,3),4),[3,1,2,4]);  % ...(_,b,c,_) ~= P( T_{i+1}^t+dt =~ c | (T_i,T_{i+1})^t = (b,c) ) / dt
        PhiT_L = kT_L.*x_RU;                  % ...(a,b,c,B) ~= P( T_{i-1}^t+dt =~ a , (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        PhiT_R = kT_R.*x_RU;                  % ...(a,b,c,B) ~= P( T_{i+1}^t+dt =~ c , (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        PhiC_C = x_RU.*permute(kC,[3,2,4,1]); % ...(a,b,c,B) =  P( C_i^t+dt =~ B     , (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        PhiT_L(isnan(PhiT_L))=0;
        PhiT_L(isinf(PhiT_L))=0;
        PhiT_R(isnan(PhiT_R))=0;
        PhiT_R(isinf(PhiT_R))=0;
        rhs = - PhiT_L + flip(PhiT_L,1)...
              - PhiT_C + flip(PhiT_C,2)...
              - PhiT_R + flip(PhiT_R,3)...
              - PhiC_C + flip(PhiC_C,4);
    end

    function x_RU_new = XB_advance(x_RU, x_XB, dt, dSLdt)
        v = -dSLdt/SL0; % [1/s]
        perm = sum(reshape(x_RU(:,2,:),[],1));
        k_PN = sum(reshape(kT(:,2,:,:).*x_RU(:,2,:,:),[],1))./sum(reshape(x_RU(:,2,:,:),[],1));
        k_NP = sum(reshape(kT(:,1,:,:).*x_RU(:,1,:,:),[],1))./sum(reshape(x_RU(:,1,:,:),[],1));
        k_PN(isnan(k_PN)) = 0; k_NP(isnan(k_NP)) = 0; % if denominator is zero, also numerator is zero
        r = r0 + alpha*abs(v);
        diag_P = r + k_PN;
        diag_N = r + k_NP;
        sol = reshape(x_XB(:,:),4,1);
        rhs = [perm*mu0_fP; perm*mu1_fP; 0; 0];
        A = [-diag_P, 0, k_NP, 0; ...
             -v, -diag_P, 0, k_NP;...
             k_PN, 0, -diag_N, 0; ...
             0, k_PN, -v, -diag_N];
        sol_inf = -A\rhs;
        x_RU_new = reshape(sol_inf + expm(dt*A)*(sol - sol_inf),[2,2]);
    end

end

File: models_python/model_RDQ20_MF.py
#!/usr/bin/env python3

# Author: Francesco Regazzoni - MOX, Politecnico di Milano
# Email:  francesco.regazzoni@polimi.it
# Date:   2020

import numpy as np
from scipy.linalg import expm
import json
import time

class model_RDQ20_MF:
    """
    Class implementing the mean-field ODE model (MF-ODE) for cardiomyocytes
    force generation presented in [1, 2].

    References
    ----------

    [1] F. Regazzoni "Mathematical modeling and Machine Learning for the
        numerical simulation of cardiac electromechanics", PhD Thesis -
        Politecnico di Milano (2020)
        http://hdl.handle.net/10589/152617
    [2] F. Regazzoni, L. Dede', A. Quarteroni "Biophysically detailed
        mathematical models of multiscale cardiac active mechanics",
        PLOS Computational Biology (2020)
        https://doi.org/10.1371/journal.pcbi.1008294
    """

    def __init__(self, params = '../params/params_RDQ20-MF_human_body-temperature.json'):
        """
        Constructor.

        Parameters
        ----------
        params : dictionary containing the model parameters, or path of a json
                 file
        """

        self.model_name = 'RDQ20-MF'

        if isinstance(params, str):
            with open(params) as json_file:
                params = json.load(json_file)

        self.LA       = float(params['geometry']['LA'])             # [micro m]
        self.LM       = float(params['geometry']['LM'])             # [micro m]
        self.LB       = float(params['geometry']['LB'])             # [micro m]
        self.SL0      = float(params['geometry']['SL0'])            # [micro m]
        self.Q        = float(params['RU_steady_state']['Q'])       # [-]
        self.Kd0      = float(params['RU_steady_state']['Kd0'])     # [micro M]
        self.alphaKd  = float(params['RU_steady_state']['alphaKd']) # [micro M / micro m]
        self.mu       = float(params['RU_steady_state']['mu'])      # [-]
        self.gamma    = float(params['RU_steady_state']['gamma'])   # [-]
        self.Koff     = float(params['RU_kinetics']['Koff'])        # [s^-1]
        self.Kbasic   = float(params['RU_kinetics']['Kbasic'])      # [s^-1]
        self.r0       = float(params['XB_cycling']['r0'])           # [s^-1]
        self.alpha    = float(params['XB_cycling']['alpha'])        # [-]
        self.mu0_fP   = float(params['XB_cycling']['mu0_fP'])       # [s^-1]
        self.mu1_fP   = float(params['XB_cycling']['mu1_fP'])       # [s^-1]
        self.a_XB     = float(params['upscaling']['a_XB'])          # [kPa]

        # Numerical parameters
        self.dt_RU = 2.5e-5                                     # [s]
        self.freqXB = round(1e-3 / self.dt_RU)                  # [-]
        self.freqenceRatesUpdate = round(2.5e-4 / 2.5e-5)       # [-]

        # Initialization
        LMh = (self.LM - self.LB) * .5;
        self.frac_SO =  lambda SL: ((SL>self.LA)          *(SL<=self.LM)          *(SL-self.LA) + \
                                    (SL>self.LM)          *(SL<=2*self.LA-self.LB)*(SL+self.LM-2*self.LA)*.5 + \
                                    (SL>2*self.LA-self.LB)*(SL<=2*self.LA+self.LB)*LMh + \
                                    (SL>2*self.LA+self.LB)*(SL<=2*self.LA+self.LM)*(self.LM+2*self.LA-SL)*.5 \
                                   )/LMh
        self.Kd = lambda SL: self.Kd0 - self.alphaKd*(2.15 - SL)
        self.kC = np.zeros((2,2))     # kC(A,a) = P( C_i^t+dt =~ A | (C_i,T_i)^t = (A,a) ) / dt
        self.kT = np.zeros((2,2,2,2)) # kT(a,b,c,B) = P( T_i^t+dt =~ b | (T_i-1,T_i,T_i+1,C_i)^t = (a,b,c,B) ) / dt
        expMat = np.array([[0, 1],[1, 2]])
        self.kC[1,0] = self.Koff
        self.kC[1,1] = self.Koff / self.mu
        self.kT[:,1,:,0] = self.Kbasic * self.gamma**(2 - expMat)
        self.kT[:,1,:,1] = self.Kbasic * self.gamma**(2 - expMat)
        self.kT[:,0,:,0] = self.Q * self.Kbasic / self.mu * self.gamma**expMat
        self.kT[:,0,:,1] = self.Q * self.Kbasic * self.gamma**expMat

    def solve(self, inputs):
        """
        Perform a simulation with the model.

        Parameters
        ----------
        inputs: dictionary containing the input data
          - inputs['times']: time instants [s]
          - inputs['Ca']:    intracellular calcium ions concentration [micro M]
          - inputs['SL']:    sarcomere length [micro m]

        Returns
        -------
        output: dictionary containing the output data
          - output['times']: time instants [s]
          - output['Ca']:    intracellular calcium ions concentration [micro M]
          - output['SL']:    sarcomere length [micro m]
          - output['P']:     permissivity [-]
          - output['Ta']:    active tension [kPa]
        """

        times = np.arange(np.min(inputs['times']), np.max(inputs['times']), self.dt_RU)
        Ca    = np.interp(times, inputs['times'], inputs['Ca'])
        SL    = np.interp(times, inputs['times'], inputs['SL'])
        nT    = len(times)
        P     = np.zeros(nT)
        Ta    = np.zeros(nT)

        old_settings = np.seterr(invalid='ignore') # ignore divide-by-zero warning
        time_init = time.time();
        print('RDQ20-MF model. Computing... ', end = '')
        for iT in range(nT):
            if iT % self.freqenceRatesUpdate == 0:
                # For performance reasons, transition rates are updated every 10 iterations
                self.__RU_update_rates(Ca[iT],SL[iT])
            if iT == 0:
                # x_RU(a,b,c,B) = P( (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) )
                x_RU = np.zeros((2,2,2,2))
                # initial state (everything in state 1, i.e. unbounded & non-permissive)
                x_RU[0,0,0,0] = 1.
                x_XB = np.zeros((2,2))
            else:
                x_RU = x_RU + self.dt_RU * self.__RU_get_rhs(x_RU)
                if iT % self.freqXB == 0:
                    dSLdt = (SL[iT] - SL[iT - self.freqXB]) / (self.dt_RU * self.freqXB)
                    x_XB = self.__XB_advance(x_RU, x_XB, self.dt_RU * self.freqXB, dSLdt)
            P[iT]  = x_RU[:,1,:,:].sum() * self.frac_SO(SL[iT])
            Ta[iT] = self.a_XB * np.sum(x_XB[1,:]) * self.frac_SO(SL[iT])
        print('done. Time elapsed: %1.3f s' % (time.time() - time_init))
        np.seterr(**old_settings)

        output = {}
        output['t']  = times
        output['Ca'] = Ca
        output['SL'] = SL
        output['P']  = P
        output['Ta'] = Ta

        return output

    def __RU_update_rates(self, Ca, SL):
        Kon = self.Koff / self.Kd(SL) # [micro M^-1 * s^-1]
        self.kC[0,:] = Kon*Ca

    def __RU_get_rhs(self, x_RU):
        PhiT_C = np.nan_to_num(x_RU*self.kT, nan = 0.0)                                # ...(a,b,c,B) =  P( T_i^t+dt =~ b, (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        kT_L   = np.sum(PhiT_C, axis = (0,3)) / np.sum(x_RU, axis = (0,3))             # ...(a,b) ~= P( T_{i-1}^t+dt =~ a | (T_{i-1},T_i)^t = (a,b) ) / dt
        kT_R   = np.sum(PhiT_C, axis = (2,3)) / np.sum(x_RU, axis = (2,3))             # ...(b,c) ~= P( T_{i+1}^t+dt =~ c | (T_i,T_{i+1})^t = (b,c) ) / dt
        PhiT_L = np.nan_to_num(kT_L[:,:,None,None]*x_RU, nan = 0.0)                    # ...(a,b,c,B) ~= P( T_{i-1}^t+dt =~ a , (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        PhiT_R = np.nan_to_num(kT_R[None,:,:,None]*x_RU, nan = 0.0)                    # ...(a,b,c,B) ~= P( T_{i+1}^t+dt =~ c , (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        PhiC_C = np.nan_to_num(x_RU * self.kC.swapaxes(1,0)[None,:,None,:], nan = 0.0) # ...(a,b,c,B) =  P( C_i^t+dt =~ B     , (T_{i-1},T_i,T_{i+1},C_i)^t = (a,b,c,B) ) / dt
        return - PhiT_L + np.flip(PhiT_L, axis = 0) \
               - PhiT_C + np.flip(PhiT_C, axis = 1) \
               - PhiT_R + np.flip(PhiT_R, axis = 2) \
               - PhiC_C + np.flip(PhiC_C, axis = 3)

    def __XB_advance(self, x_RU, x_XB, dt, dSLdt):
        v = -dSLdt / self.SL0 # [1/s]
        perm = x_RU[:,1,:,:].sum()
        k_PN = np.nan_to_num((self.kT[:,1,:,:] * x_RU[:,1,:,:]).sum() / x_RU[:,1,:,:].sum(), nan = 0.0)
        k_NP = np.nan_to_num((self.kT[:,0,:,:] * x_RU[:,0,:,:]).sum() / x_RU[:,0,:,:].sum(), nan = 0.0)
        r = self.r0 + self.alpha*abs(v)
        diag_P = r + k_PN
        diag_N = r + k_NP
        sol = np.reshape(x_XB, (4,), order='F')
        rhs = np.array((perm*self.mu0_fP, perm*self.mu1_fP, 0.0, 0.0))
        A = np.array([[-diag_P, 0      , k_NP   , 0      ], \
                      [-v     , -diag_P, 0      , k_NP   ], \
                      [k_PN   , 0      , -diag_N, 0      ], \
                      [0      , k_PN   , -v     , -diag_N]])
        sol_inf = -np.linalg.solve(A, rhs)
        return np.reshape(sol_inf + np.matmul(expm(dt*A), sol - sol_inf), (2,2), order='F')